### PR TITLE
V628-024 Symbol diff has been added

### PR DESCRIPTION
**Pull request by ogorodnik**

### CODE REVIEW

File: source/ada/lsp-ada_documents.adb [hunk 1]
```diff
@@ -715,19 +1017,37 @@ package body LSP.Ada_Documents is
...
 
          Out_Span := Self.To_LSP_Range (Out_Sloc);
-         Diff
-           (Self,
-            VSS.Strings.Conversions.To_Virtual_String (S.all),
-            Span,
-            Out_Span,
-            Edit);
+
+         if Span.last.line - Span.first.line > 5 then
```
Comment (AnthonyLeonardoGracio) [inline source/ada/lsp-ada_documents.adb:1021]: Need a comment explaining this threshold

**Review by AnthonyLeonardoGracio — CHANGES_REQUESTED** [review]
I'll let Vadim comment on the VSS part in itself, since I am not an expert. We miss some tests too

**Review by godunko — DISMISSED** [review]
Looks good

File: source/ada/lsp-ada_documents.adb (hunk 1 above)
Comment (reznikmm) [reply source/ada/lsp-ada_documents.adb:1021]: You have a space at the end of comment, I guess

File: source/ada/lsp-ada_documents.adb [hunk 2]
```diff
@@ -604,6 +605,307 @@ package body LSP.Ada_Documents is
...
+         --  to calculate span
+         Current_Line_Number : Line_Number := Span.last.line;
+         Old_Lines_Number    : Natural := Old_Lines.Length;
+         Cursor              : VSS.Strings.Cursors.Iterators.Characters.
+           Character_Iterator := Old_Line.After_Last_Character;
+
+         procedure Backward;
+         --  Move old line "cursors" backward
+
+         function Get_Span (Insert : Boolean) return Position;
```
[the remark this answers is not in the document]
Comment (ogorodnik) [reply source/ada/lsp-ada_documents.adb]: Done

File: source/ada/lsp-ada_documents.adb [hunk 3]
```diff
@@ -604,6 +605,307 @@ package body LSP.Ada_Documents is
...
+         --  get Position for a Span
+
+         procedure Prepare (Insert : Boolean);
+         procedure Prepare
+           (Insert : Boolean;
+            Char   : VSS.Characters.Virtual_Character);
+         --  Store imformation for Text_Etid in New_String and Span
```
[the remark this answers is not in the document]
Comment (ogorodnik) [reply source/ada/lsp-ada_documents.adb]: Done

File: source/ada/lsp-ada_documents.adb [hunk 4]
```diff
@@ -604,6 +605,314 @@ package body LSP.Ada_Documents is
...
+           (Insert : Boolean;
+            Char   : VSS.Characters.Virtual_Character);
+         --  Store information for Text_Edit in Changed_Block_Text and
+         --  Changed_Block_Span
+
+         procedure Add_Prepared;
+         --  Add prepared New_String and Span into Text_Edit
+
+         procedure Backward is
+         begin
```
Comment (AnthonyLeonardoGracio) [inline source/ada/lsp-ada_documents.adb:699]: Missing subprogram box

File: source/ada/lsp-ada_documents.adb [hunk 5]
```diff
@@ -604,6 +605,314 @@ package body LSP.Ada_Documents is
...
+         --  to calculate span
+         Current_Line_Number : Line_Number := Span.last.line;
+         Old_Lines_Number    : Natural := Old_Lines.Length;
+         Cursor              : VSS.Strings.Cursors.Iterators.Characters.
+           Character_Iterator := Old_Line.After_Last_Character;
+
+         procedure Backward;
+         --  Move old line "cursors" backward
+
+         function Get_Position (Insert : Boolean) return Position;
```
Comment (AnthonyLeonardoGracio) [inline source/ada/lsp-ada_documents.adb:685]: Which position? For which span? The comment is not clear at all to me

File: source/ada/lsp-ada_documents.adb [hunk 6]
```diff
@@ -604,6 +605,314 @@ package body LSP.Ada_Documents is
...
+            -- Backward --
+            --------------
+
+            function Backward return Position;
+            function Backward return Position is
+               C : VSS.Strings.Cursors.Iterators.Characters.
+                 Character_Iterator := Old_Line.At_Character (Cursor);
+            begin
+               --  "Cursor" is after the current character but we should
+               --  inserting before it
```
Comment (AnthonyLeonardoGracio) [inline source/ada/lsp-ada_documents.adb:729]: "insert before it"

File: source/ada/lsp-ada_documents.adb [hunk 7]
```diff
@@ -604,6 +605,314 @@ package body LSP.Ada_Documents is
...
+
+         begin
+            if not Cursor.Has_Element then
+               return
+                 (line      => Current_Line_Number,
+                  character => 0);
+
+            elsif Insert then
+               --  "Cursor" is after the current character but we should
+               --  inserting before it
```
Comment (AnthonyLeonardoGracio) [inline source/ada/lsp-ada_documents.adb:749]: "insert before it"

File: source/ada/lsp-ada_documents.adb [hunk 8]
```diff
@@ -604,6 +605,314 @@ package body LSP.Ada_Documents is
...
+               Span_Set := True;
+               Changed_Block_Span.last := Get_Position (Insert);
+            end if;
+         end Prepare;
+
+         -------------
+         -- Prepare --
+         -------------
+
+         procedure Prepare
```
Comment (AnthonyLeonardoGracio) [inline source/ada/lsp-ada_documents.adb:777]: Can't we have a more precise name than 'Prepare'?

**Review by AnthonyLeonardoGracio — COMMENTED** [review]
The code is really hard to read IMO. Maybe it's juste me, but I think you can try to improve a bit more the comments and the function names if other people need to resolve a bug later in that area.

File: source/ada/lsp-ada_documents.adb (hunk 4 above)
Comment (ogorodnik) [reply source/ada/lsp-ada_documents.adb:708]: Done

File: source/ada/lsp-ada_documents.adb (hunk 5 above)
Comment (ogorodnik) [reply source/ada/lsp-ada_documents.adb:686]: Done

File: source/ada/lsp-ada_documents.adb (hunk 6 above)
Comment (ogorodnik) [reply source/ada/lsp-ada_documents.adb]: +

File: source/ada/lsp-ada_documents.adb (hunk 7 above)
Comment (ogorodnik) [reply source/ada/lsp-ada_documents.adb]: Done

File: source/ada/lsp-ada_documents.adb (hunk 8 above)
Comment (ogorodnik) [reply source/ada/lsp-ada_documents.adb]: Changed

**Pull request closed without merging**